pep-3157.txt

```diff
@@ -0,0 +1,247 @@
+
+    >>> f'''A complex trick: {
+    ... bag['bag']  # recursive bags!
+    ... }'''
+    SyntaxError: f-string expression part cannot include '#'
+
+#. (TODO: DO WE WANT THIS?) Expression portions need to wrap ``':'`` and ``'!'`` in braces::
+
+    >>> f'Useless use of lambdas: { lambda x: x*2 }'
+    SyntaxError: unexpected EOF while parsing
+
+#. Arbitrary nesting of expressions without expansion of escape sequences is
+   available in every single other language employing a string interpolation
+   method that uses expressions instead of just variable names. [6]_
+
+These limitations serve no purpose from a language user perspective and
+can be lifted by giving f-literals a regular grammar without exceptions
+and implementing it using dedicated parse code.
+
+The other problems that f-strings have is that the current implementation in
+CPython relies on the existing string parsing machinery and a post processing of
+its tokens. This has the following problems:
+
+#. It adds a considerable maintainance cost to the CPython parser. This is because
+   the parsing code needs to be written by hand, which has historically lead to a
+   considerable number of inconsistencies and bugs. Writing and maintaining parsing
+   code by hand in C has always been considered error prone and dangerous as it needs
+   to deal with a lot of manual memory management over the original lexer buffers.
+
+#. The f-string parsing code is not able to use the new improved error message mechanisms
+   that the new PEG parser originally introduced in :pep:`617` has allowed. The
+   improvements that these error messages has been greatly celebrated but
+   unfortunately f-strings cannot benefit from them because they are parser in a
+   separate piece of the parsing machinery. This is specially unfortunate, since
+   there are several syntactical features of f-strings that can be confusing due
+   to the different implicit tokenization that happens inside the expression
+   part (for instance ``f"{y:=3}"`` is not an assignment expression).
+
+#. Other Python implementations have no way to know if they have implemented
+   f-strings correctly because contrary to other language features, they are not
+   part of the official Grammar [1]_. This is important because several prominent
+   alternative implementations such as PyPy are using CPython's PEG parser
+   (see [3]_) and/or are basing their Grammars on the official PEG Grammar. The
+   fact that f-strings use a separate parser prevents these alternative implementations
+   to leverage the official grammar and to benefit to improvements in error messages derived
+   from the grammar.
+
+
+A version of this proposal was originally discussed in Python dev [4]_ and
+presented at the Python language summit 2022 [5]_ where it was enthusiastically
+received.
+
+Rationale
+=========
+
+[Describe why particular design decisions were made.]
+
+TO BE DONE:
+
+* Describe why we are allowing crazy parenthesis nesting.
```

Comment:
@pablogsal Do you mean crazy f-string nesting here?

Reply:
Yeah, apologies. I meant why we are allowing crazy "nest f-strings with the same quote"